### DISCUSSION TITLE
Use HashSet rather than string[] for songScene, and similar lists

### DISCUSSION THREAD

**Discussion by kozaka-tv** [Th0rstenf/Rocksniffer-to-Streamer.bot-extension#91] [Ideas]
And then use HashSet<T>.Contains(T) method, to get for example the information like IsSongScene.
This way code could be a bit simpler compared to the foreach solution.

This code

![image](https://user-images.githubusercontent.com/4964462/230747786-0ab924db-1cff-4633-b90f-683d4b3827a8.png)

 could be replaced with this:

![image](https://user-images.githubusercontent.com/4964462/230747842-1f5cec40-e5b7-4fbc-9ceb-e337de06efb1.png)

The LINQ-expression solution is not working in SB because the shit older .Net FW version. And the problem is, actually, this would be the best. Simple to read, and using a performant array.

![image](https://user-images.githubusercontent.com/4964462/231856256-dfb1ccec-ec5a-4ae6-898b-94d68ed6cfd8.png)

**Comment by Th0rstenf** [comment]
A hash set has constant times for everything, but far larger overhead. For any small set and we're rather talking TINY than small, the current solution is much more performant. 

That being said, it could still be changed, but your reason is not really correct.
Do you see other advantages?

While I'm usually all for performance, it's really not relevant. If we check execution times of the whole thing, we're still hardly noticeable in load. So if there are other benefits in totally OK with sacrificing performance for other benefits

However those should be significant. If we think further and we want to access a random element from the list, for a random scene change later, it will be even worse in performance ;)

**Comment by kozaka-tv** [comment]
Idk...maybe it does not matter, but at least, code would be simpler and more error free. We could have for example except this: 

![image](https://user-images.githubusercontent.com/4964462/230747786-0ab924db-1cff-4633-b90f-683d4b3827a8.png)

A way more simpler like this:

![image](https://user-images.githubusercontent.com/4964462/230747842-1f5cec40-e5b7-4fbc-9ceb-e337de06efb1.png)

I am not sure how this set stuff in c# working, but could not imagine, that there would be so much difference in our case, as the set's contains method probably using some index, and it is then fast, and created only once, in Init. To run over a list always, should be more costly. And we do not change those lists, so no add or delete happens, just a search with contains normally.

I had to read a bit, but this, if I unterstand correctly, it shows, that set's, regarding search (contains) are more effective than lists: https://stackoverflow.com/a/13089134/3462911 but they are like very-very the same. In our case, the list is like from 1 to max 5 for the songScenes...or even 90% of the user will have only one. Maybe you red the same article, but the main answer, answers not the question. This is so bad...question was regarding search only. And the answer with 1000+ likes is showing a different stuff.

For me, less code is better. So using a HashSet could help in this. We will do not see any performance change I think.

**Reply by Th0rstenf** [reply]
> I am not sure how this set stuff in c# working, but could not imagine, that there would be so much difference in our case, as the set's contains method probably using some index, and it is then fast, and created only once, in Init. To run over a list always, should be more costly. And we do not change those lists, so no add or delete happens, just a search with contains normally.

> I had to read a bit, but this, if I unterstand correctly, it shows, that set's, regarding search (contains) are more effective than lists: https://stackoverflow.com/a/13089134/3462911 but they are like very-very the same. In our case, the list is like from 1 to max 5 for the songScenes...or even 90% of the user will have only one. Maybe you red the same article, but the main answer, answers not the question. This is so bad...question was regarding search only. And the answer with 1000+ likes is showing a different 

A hashset, as the name implies  calculates a hash value for the checks, you can probably iterate over 100 elements before this is done. Again, its main advantage is deterministic CPU cycles, not few.

The article compares lists to hashsets. Lists already are significantly worse in performance compared to raw arrays. They both have other advantages, but in terms of performance they are not even close to being competitive for small sets (and small in that case refers to couple of hundreds rather than dozens of elements) List would also be horrible for anything other than sequential access.

Those are not arguments against the change, I just want to understand this, because stating a hashset is faster than an array of less than a dozen elements is far from realistic.

We should check first if there are no better methods to work with the array of if there is an array class? c++ would offer vector or array class templates that would be perfect.
Especially considering that sequential is a use case as well, array/list seem like logical choices but with list, random access is not very convenient

> For me, less code is better. So using a HashSet could help in this. We will do not see any performance change I think.

Less code CAN be better, it also can be worse. Especially when it comes to debugging. But using more library methods will make it safer to use, and that's a real plus. I'm still not convinced which container is the best, considering the use cases that are already in the backlog.

We should really set our target to 4.0 then though, otherwise the probability of different behavior is dramatically increased.

**Comment by Th0rstenf** [comment]
And just reading your article, there are already comments stating that for such small numbers, lists easily outperform the hash set if size is specified.
And that would still be slower than any array

**Reply by Th0rstenf** [reply]
My suggestion would be sticking to array, but instead of comparing each  entry, we use find with a predicate. Then we have less code, that probably does the same thing internally

**Reply by kozaka-tv** [reply]
Yes, but there is the problem, so far I understud, that the question was about search (random search), and not add/delete. Even the question starts like, List is faster...what I do not understand, why he is asking, if he knows the answer :) and the answer with the most votes, meassure the add/delete. What is again not the question it originally was.

I mean, our list are very small, my idea to use Hashset was because it is faster was not correct or actually irrelevant. I thought, that the list is created at every run. What I wanted to suggest, was that, to run over a List always to find an element, or get it from a set, then set is better. No duplicates, and probably same foreach stuff is used, or hash...idk. The lists are not like having 10000 elements, so just for us, the code could be better if simpler...that's what I ment.

As I am a Java guy, I can just know how it there works, and not with c#. To change from a list to set, makes probably no performance change at all. I just do not like the foreach each time I want to check, a list.

**Reply by kozaka-tv** [reply]
But I checked HashSet.Contains...man this is complex.

So yeah...predicate could be a good solution to use, for better code and performance. And keep the arrays.

**Reply by Th0rstenf** [reply]
I think that's what we'll do. By the way, the way this works is independent of the language, you just need to understand the algorithm and think what that relates to in instructions

**Reply by Th0rstenf** [reply]
Oh, another disadvantage. Users are ALLOWED to have duplicates. e.g. they consider a sequential number of scenes, but want every second scene to be a specific one.

**Reply by Th0rstenf** [reply]
this is what the predicate would look like:

``` 
private bool IsSongScene(string scene)
{
    return Array.Find<string>(songScenes, s => s.Equals(scene)) != null;
}
```


I honesty can't say that it's superior to the original in terms of readability (most likely also not in performance). Less code is not always better.
I'll still add this minor change to the current pull request

**Reply by kozaka-tv** [reply]
What's about Lists? Not string[] rather than we could use a List.

In init we create a List... then:  songScenes.Contains(scene);

**Reply by Th0rstenf** [reply]
Lists would be ok, but honestly the array is better.
Find with predicate delivers a fast and elegant lookup option. We have ideal sequential access, and can access random elements at deterministic (and minimal) time.
You already said it. We create it once at init. There will neither be objects removed nor added during runtime, which would be the main reason for chosing a List. If the amount of elements is known to be constants, List are never the ideal option

**Reply by kozaka-tv** [reply]
Then the foreach solution could stay. Or when, then a List would be enough now, because of the current .NET Version. If we use List, then the .Contains() could be used.

So, not so important to change anything.
